### ISSUE TITLE
Generational backups not working

### ISSUE THREAD

**Issue by Jurgen-DOUCHY** [lavinir/hassio-onedrive-backup#205]
Ik looks like generational backups are not working.

I have my backups on last 7 daily + last 4 weekly + 12 monthly and 10 yearly
![image](https://github.com/lavinir/hassio-onedrive-backup/assets/120566296/332b65de-6b63-4f24-92f2-86becae348fa)

But all my backups are saved.
It looks like the settings are ignored.
![image](https://github.com/lavinir/hassio-onedrive-backup/assets/120566296/513ebe63-04de-45da-ac81-3442ab445726)

Am I missing something?

**Comment by lloyda** [comment]
I'm seeing the same thing.

**Comment by lloyda** [comment]
From the logs, it looks like the correct backups to be deleted have been identified, but they are never deleted.

```
05/09/2024 16:27:26 [43] Verbose: File Sync Disabled...
05/09/2024 16:27:52 [18] Verbose: Checking backups
05/09/2024 16:27:52 [18] Verbose: Refreshing existing backups...
05/09/2024 16:27:52 [18] Verbose: Backup Interval Completed.
05/09/2024 16:27:52 [18] Verbose: Ignoring 'External' backup: WeeklyBackup: Monday, April 15, 2024
05/09/2024 16:27:52 [18] Verbose: Ignoring 'External' backup: MonthlyBackup: Monday, April 1, 2024
05/09/2024 16:27:52 [18] Verbose: Ignoring 'External' backup: MonthlyBackup: Wednesday, November 1, 2023
05/09/2024 16:27:52 [18] Verbose: Ignoring 'External' backup: core_2024.4.1
05/09/2024 16:27:52 [18] Verbose: Ignoring 'External' backup: MonthlyBackup: Tuesday, August 1, 2023
05/09/2024 16:27:52 [18] Verbose: Ignoring 'External' backup: MonthlyBackup: Thursday, June 1, 2023
05/09/2024 16:27:52 [18] Verbose: Ignoring 'External' backup: MonthlyBackup: Monday, May 1, 2023
05/09/2024 16:27:52 [18] Verbose: Ignoring 'External' backup: WeeklyBackup: Monday, March 25, 2024
05/09/2024 16:27:52 [18] Verbose: Ignoring 'External' backup: MonthlyBackup: Friday, March 1, 2024
05/09/2024 16:27:52 [18] Verbose: Ignoring 'External' backup: core_2024.5.1
05/09/2024 16:27:52 [18] Verbose: Ignoring 'External' backup: WeeklyBackup: Monday, April 8, 2024
05/09/2024 16:27:52 [18] Verbose: Ignoring 'External' backup: core_2024.4.3
05/09/2024 16:27:52 [18] Verbose: Ignoring 'External' backup: core_2024.4.4
05/09/2024 16:27:52 [18] Verbose: Ignoring 'External' backup: MonthlyBackup: Sunday, October 1, 2023
05/09/2024 16:27:52 [18] Verbose: Ignoring 'External' backup: YearlyBackup: Sunday, January 1, 2023
05/09/2024 16:27:52 [18] Verbose: Ignoring 'External' backup: core_2024.3.3
05/09/2024 16:27:52 [18] Verbose: Ignoring 'External' backup: Pre ramses_cc upgrade
05/09/2024 16:27:52 [18] Verbose: Ignoring 'External' backup: MonthlyBackup: Friday, December 1, 2023
05/09/2024 16:27:52 [18] Verbose: Ignoring 'External' backup: addon_core_ssh_9.13.0
05/09/2024 16:27:52 [18] Verbose: Ignoring 'External' backup: core_2024.4.2
05/09/2024 16:27:52 [18] Verbose: Ignoring 'External' backup: YearlyBackup: Monday, January 1, 2024
05/09/2024 16:27:52 [18] Verbose: Ignoring 'External' backup: MonthlyBackup: Saturday, July 1, 2023
05/09/2024 16:27:52 [18] Verbose: Ignoring 'External' backup: addon_a0d7b954_grafana_9.2.1
05/09/2024 16:27:52 [18] Verbose: Ignoring 'External' backup: MonthlyBackup: Thursday, February 1, 2024
05/09/2024 16:27:52 [18] Verbose: Ignoring 'External' backup: MonthlyBackup: Friday, September 1, 2023
05/09/2024 16:27:53 [38] Verbose: Last local backup Date: 05/09/2024 01:09:09
05/09/2024 16:27:53 [38] Verbose: Last online backup Date: 05/09/2024 01:09:09
05/09/2024 16:27:53 [38] Verbose: Online backups synced. No upload required
05/09/2024 16:27:53 [38] Verbose: Evaluating Online Generational Backups
05/09/2024 16:27:53 [38] Verbose: Backup (b83ae5a1 (05/04/2024 01:00:41) retained for Generational Daily policy OneDrive
05/09/2024 16:27:53 [38] Verbose: Backup (b1124422 (05/05/2024 01:01:12) retained for Generational Daily policy OneDrive
05/09/2024 16:27:53 [38] Verbose: Backup (228dc4ff (05/06/2024 01:01:45) retained for Generational Daily policy OneDrive
05/09/2024 16:27:53 [38] Verbose: Backup (b219d659 (05/07/2024 01:02:55) retained for Generational Daily policy OneDrive
05/09/2024 16:27:53 [38] Verbose: Backup (b22c747c (05/08/2024 01:05:36) retained for Generational Daily policy OneDrive
05/09/2024 16:27:53 [38] Verbose: Backup (fddb5250 (05/09/2024 01:09:09) retained for Generational Daily policy OneDrive
05/09/2024 16:27:53 [38] Verbose: Backup (fddb5250 (05/09/2024 01:09:09) retained for Generational Weekly policy OneDrive
05/09/2024 16:27:53 [38] Verbose: Backup (b83ae5a1 (05/04/2024 01:00:41) retained for Generational Weekly policy OneDrive
05/09/2024 16:27:53 [38] Verbose: Backup (3e0bcbe2 (04/27/2024 01:16:01) retained for Generational Weekly policy OneDrive
05/09/2024 16:27:53 [38] Verbose: Backup (af515d34 (04/20/2024 01:07:50) retained for Generational Weekly policy OneDrive
05/09/2024 16:27:53 [38] Verbose: Backup (fddb5250 (05/09/2024 01:09:09) retained for Generational Monthly policy OneDrive
05/09/2024 16:27:53 [38] Verbose: Backup (9960f4f6 (04/30/2024 01:18:06) retained for Generational Monthly policy OneDrive
05/09/2024 16:27:53 [38] Verbose: Backup (fddb5250 (05/09/2024 01:09:09) retained for Generational Yearly policy OneDrive
05/09/2024 16:27:53 [38] Verbose: Found 15 OneDrive backups that can be removed (Generational Rules)
05/09/2024 16:27:53 [38] Verbose: Potential OneDrive backups for removal: 3022b976 (04/14/2024 10:21:20),8f7c5796 (04/16/2024 01:00:46),b33e53f3 (04/17/2024 01:02:47),9d3be703 (04/18/2024 01:04:41),162f3f8a (04/19/2024 01:06:36),79db235d (04/21/2024 01:08:28),656873a7 (04/22/2024 01:09:14),529b8d58 (04/23/2024 01:10:21),7befcceb (04/24/2024 01:11:58),a10139cd (04/25/2024 01:13:38),7e1429d3 (04/26/2024 01:14:31),634518ef (04/28/2024 01:16:32),ac1fb793 (04/29/2024 01:16:53),5dde4913 (05/01/2024 01:19:53),20f4401b (05/02/2024 17:50:20)
05/09/2024 16:27:53 [38] Verbose: Evaluating Local Generational Backups
05/09/2024 16:27:53 [38] Verbose: Backup (b219d659 (05/07/2024 01:02:55) retained for Generational Daily policy local
05/09/2024 16:27:53 [38] Verbose: Backup (228dc4ff (05/06/2024 01:01:45) retained for Generational Daily policy local
05/09/2024 16:27:53 [38] Verbose: Backup (b22c747c (05/08/2024 01:05:36) retained for Generational Daily policy local
05/09/2024 16:27:53 [38] Verbose: Backup (b83ae5a1 (05/04/2024 01:00:41) retained for Generational Daily policy local
05/09/2024 16:27:53 [38] Verbose: Backup (fddb5250 (05/09/2024 01:09:09) retained for Generational Daily policy local
05/09/2024 16:27:53 [38] Verbose: Backup (b1124422 (05/05/2024 01:01:12) retained for Generational Daily policy local
05/09/2024 16:27:53 [38] Verbose: Backup (fddb5250 (05/09/2024 01:09:09) retained for Generational Weekly policy local
05/09/2024 16:27:53 [38] Verbose: Backup (b83ae5a1 (05/04/2024 01:00:41) retained for Generational Weekly policy local
05/09/2024 16:27:53 [38] Verbose: Backup (3e0bcbe2 (04/27/2024 01:16:01) retained for Generational Weekly policy local
05/09/2024 16:27:53 [38] Verbose: Backup (af515d34 (04/20/2024 01:07:50) retained for Generational Weekly policy local
05/09/2024 16:27:53 [38] Verbose: Backup (fddb5250 (05/09/2024 01:09:09) retained for Generational Monthly policy local
05/09/2024 16:27:53 [38] Verbose: Backup (9960f4f6 (04/30/2024 01:18:06) retained for Generational Monthly policy local
05/09/2024 16:27:53 [38] Verbose: Backup (fddb5250 (05/09/2024 01:09:09) retained for Generational Yearly policy local
05/09/2024 16:27:53 [38] Verbose: Found 15 local backups that can be removed (Generational Rules)
05/09/2024 16:27:53 [38] Verbose: Potential local backups for removal: 20f4401b (05/02/2024 17:50:20),9d3be703 (04/18/2024 01:04:41),656873a7 (04/22/2024 01:09:14),3022b976 (04/14/2024 10:21:20),79db235d (04/21/2024 01:08:28),7e1429d3 (04/26/2024 01:14:31),5dde4913 (05/01/2024 01:19:53),a10139cd (04/25/2024 01:13:38),ac1fb793 (04/29/2024 01:16:53),529b8d58 (04/23/2024 01:10:21),b33e53f3 (04/17/2024 01:02:47),7befcceb (04/24/2024 01:11:58),634518ef (04/28/2024 01:16:32),8f7c5796 (04/16/2024 01:00:46),162f3f8a (04/19/2024 01:06:36)
05/09/2024 16:27:53 [38] Verbose: Ignoring 'External' backup: WeeklyBackup: Monday, April 15, 2024
05/09/2024 16:27:53 [38] Verbose: Ignoring 'External' backup: MonthlyBackup: Monday, April 1, 2024
05/09/2024 16:27:53 [38] Verbose: Ignoring 'External' backup: MonthlyBackup: Wednesday, November 1, 2023
05/09/2024 16:27:53 [38] Verbose: Ignoring 'External' backup: core_2024.4.1
05/09/2024 16:27:53 [38] Verbose: Ignoring 'External' backup: MonthlyBackup: Tuesday, August 1, 2023
05/09/2024 16:27:53 [38] Verbose: Ignoring 'External' backup: MonthlyBackup: Thursday, June 1, 2023
05/09/2024 16:27:53 [38] Verbose: Ignoring 'External' backup: MonthlyBackup: Monday, May 1, 2023
05/09/2024 16:27:53 [38] Verbose: Ignoring 'External' backup: WeeklyBackup: Monday, March 25, 2024
05/09/2024 16:27:53 [38] Verbose: Ignoring 'External' backup: MonthlyBackup: Friday, March 1, 2024
05/09/2024 16:27:53 [38] Verbose: Ignoring 'External' backup: core_2024.5.1
05/09/2024 16:27:53 [38] Verbose: Ignoring 'External' backup: WeeklyBackup: Monday, April 8, 2024
05/09/2024 16:27:53 [38] Verbose: Ignoring 'External' backup: core_2024.4.3
05/09/2024 16:27:53 [38] Verbose: Ignoring 'External' backup: core_2024.4.4
05/09/2024 16:27:53 [38] Verbose: Ignoring 'External' backup: MonthlyBackup: Sunday, October 1, 2023
05/09/2024 16:27:53 [38] Verbose: Ignoring 'External' backup: YearlyBackup: Sunday, January 1, 2023
05/09/2024 16:27:53 [38] Verbose: Ignoring 'External' backup: core_2024.3.3
05/09/2024 16:27:53 [38] Verbose: Ignoring 'External' backup: Pre ramses_cc upgrade
05/09/2024 16:27:53 [38] Verbose: Ignoring 'External' backup: MonthlyBackup: Friday, December 1, 2023
05/09/2024 16:27:53 [38] Verbose: Ignoring 'External' backup: addon_core_ssh_9.13.0
05/09/2024 16:27:53 [38] Verbose: Ignoring 'External' backup: core_2024.4.2
05/09/2024 16:27:53 [38] Verbose: Ignoring 'External' backup: YearlyBackup: Monday, January 1, 2024
05/09/2024 16:27:53 [38] Verbose: Ignoring 'External' backup: MonthlyBackup: Saturday, July 1, 2023
05/09/2024 16:27:53 [38] Verbose: Ignoring 'External' backup: addon_a0d7b954_grafana_9.2.1
05/09/2024 16:27:53 [38] Verbose: Ignoring 'External' backup: MonthlyBackup: Thursday, February 1, 2024
05/09/2024 16:27:53 [38] Verbose: Ignoring 'External' backup: MonthlyBackup: Friday, September 1, 2023

**Comment by lavinir** [comment]
The generational deletions are non-aggressive, so if the number of backups are underneath the Max Backups in your config they will not be deleted. Once the Max Backups number is reached, the backups will be deleted according to the set generational rules.

**Comment by lavinir** [comment]
I think that introduces complications to the logic. The generational check currently only applies to retention not the backup creation schedule. 
Say for example I have configured 1 daily and 2 weekly backups and have set my backup schedule to daily.
Every day when a new backup is created, based on retention only it will delete the previous days backup, essentially preventing a backup to be kept for over a week.

I tried going that route initially when I was implementing this but it introduced too many cases to handle and potential breaks.
